feat: default new repos to CIDv1

lidel opened the pull request:
## Problem

Making CIDv1 (base32 `bafy...`) the default has been tracked in #4143 for years, and the groundwork already landed: both CID versions work throughout, the CLI and gateway render base32, and Kubo 0.29's `Import` config (#10421) together with [IPIP-499](https://specs.ipfs.tech/ipips/ipip-0499/)'s `unixfs-v0-2015` / `unixfs-v1-2025` profiles pin the parameters so any implementation can reproduce the same CIDs. The one remaining gap was the default itself: `ipfs init` wrote no `Import.*`, so new repos fell back to CIDv0 `Qm...`.

Some users also lean on CIDs as if they were deterministic hashes. That is a misunderstanding of what a CID is, but breaking them isn't the goal, so existing repos need agency to move to CIDv1 on their own schedule.

## Fix

`ipfs init` now applies the `unixfs-v1-2025` profile to the config it writes, so a fresh repo produces CIDv1 everywhere. This is deliberately the least disruptive way to flip the default:

- **Only fresh repos change.** The switch lives in what `ipfs init` writes, not a code-level fallback. A pre-existing repo with empty or absent `Import.*` still resolves to CIDv0 exactly as before, with no migration and no repo-version bump.
- **Every import path follows the repo's config,** not just `ipfs add`: `ipfs files`, the MFS root, the gateway, and the `client/rpc` Go library all produce the configured CID version. Requesting a version explicitly (`--cid-version`, or a non-sha2-256 hash) uses that version's natural leaves; contradictions like `--cid-version=0 --raw-leaves`, or a config pairing `Import.CidVersion=0` with `Import.UnixFSRawLeaves=true`, are rejected rather than emitting a surprising CID.
- **`Import.*` is the single source of truth,** including whether the startup notice appears. If any CID-profile field is still unset, `ipfs daemon` prints a one-time notice nudging the operator to pin a profile; once explicit, it stops.
- **Users choose, explicitly.** New repos get `unixfs-v1-2025`; `ipfs init --profile unixfs-v0-2015` keeps CIDv0; existing-repo operators are nudged to record their choice rather than lean on an implicit default that could shift across releases.

This respects user agency and avoids breaking legacy datasets, while making CIDv1 what new users get out of the box.

## References

- Context
  - **Spec:** [IPIP-499: UnixFS CID Profiles](https://specs.ipfs.tech/ipips/ipip-0499/) ([ipfs/specs#499](https://github.com/ipfs/specs/pull/499)) defines the `unixfs-v0-2015` and `unixfs-v1-2025` profiles this PR applies.
  - **Enabling work:** [#10421](https://github.com/ipfs/kubo/pull/10421) added the `Import` config section; [#11148](https://github.com/ipfs/kubo/pull/11148) added its IPIP-499 profile fields.
  - **Groundwork:** defaulting to CIDv1 became practical once base32 ([#6220](https://github.com/ipfs/kubo/issues/6220)) and subdomain gateways ([#6498](https://github.com/ipfs/kubo/issues/6498)) shipped.
  - **Supersedes:** [#8886](https://github.com/ipfs/kubo/pull/8886), [#8896](https://github.com/ipfs/kubo/pull/8896), and [#8898](https://github.com/ipfs/kubo/pull/8898), three earlier attempts at CIDv1-by-default.
  - **Interop:** [ipfs/helia#1083](https://github.com/ipfs/helia/pull/1083) updates the Helia interop suite for the new default.
- Closes #4143
- Closes ipfs/specs#247

## Demo

**A fresh repo now defaults to CIDv1.** `ipfs init` writes the `unixfs-v1-2025` profile, so `ipfs add`, `ipfs files`, and the MFS root produce base32 `bafy.../bafk...` out of the box instead of `Qm...`:

```console
$ ipfs init          # writes the unixfs-v1-2025 profile
$ ipfs config Import.CidVersion
1
$ echo -n "hello, CIDv1!" | ipfs add -q
bafkreifriwiigahgxhtea4usek2ew2hr4vtnzsj7m6jy6n5iusfporvs2e   # was QmNbyD5G... before
```

**Existing repos are untouched.** A repo whose `Import.*` is empty or absent keeps producing the exact same CIDs. The only new thing its operator sees is a one-time notice at `ipfs daemon` startup, until a profile is pinned:

```console
$ ipfs daemon
Initializing daemon...
...
⚠️ This repository relies on implicit UnixFS Import.* config defaults,
⚠️ so the CIDs it produces can change between Kubo releases. Pin
⚠️ them by applying a profile (this also silences this notice):
⚠️   $ ipfs config profile apply unixfs-v1-2025   # recommended, modern CIDv1
⚠️   $ ipfs config profile apply unixfs-v0-2015   # legacy, CIDv0 "Qm.." format
⚠️ Use the legacy profile only to keep reproducing existing CIDs.
⚠️ See IPIP-499: https://specs.ipfs.tech/ipips/ipip-0499/
...
Daemon is ready
```

Once the operator applies a profile (or runs `ipfs init --profile ...` on a new repo), `Import.*` is explicit and the notice stops.


## TODO

- [x] land https://github.com/ipfs/helia/pull/1083 first

schomatis commented:
Similar to https://github.com/ipfs/go-ipfs/pull/8886, leaving sharness CID translation for last.

lidel committed:
feat: default new repos to CIDv1

New `ipfs init` applies the unixfs-v1-2025 (IPIP-499) import profile, so
fresh repositories produce CIDv1 base32 instead of CIDv0. Existing repos
are untouched: no migration and no repo-version bump, so upgraded nodes
keep producing the same CIDs.

- config: profiles are named via ProfileUnixFSv12025 / ProfileUnixFSv02015;
  the values used when an Import.* field is unset are LegacyFallback*
  consts, kept at CIDv0 so existing repos are unchanged and the name marks
  them as migration-only fallbacks
- fresh repos are CIDv1 end to end: MFS root, self-IPNS empty dir, and
  seeded init docs follow the configured import profile
- UnixfsAPI.Add honors the node's Import config, so every import path
  (direct callers, gateway, client/rpc) produces the repo's default CID
  version, not just `ipfs add`
- CidVersion and raw leaves resolve as a bundle: the config's raw-leaves
  default applies only when the effective CID version matches the config's,
  so requesting a different version with --cid-version or a non-sha2-256
  hash uses that version's natural leaves (CIDv1 raw, CIDv0 dag-pb);
  --cid-version=0 with --raw-leaves, or a config pairing CidVersion=0 with
  UnixFSRawLeaves=true, is rejected
- client/rpc: Add follows the daemon's configured CID version instead of
  always sending CIDv0, and forwards the max-links and HAMT-fanout options
- daemon: warn once when a repo still relies on implicit UnixFS import
  defaults, pointing to the two IPIP-499 profiles
- legacy sharness pins the unixfs-v0-2015 profile to keep its CIDv0 fixtures

Closes #4143

lidel committed:
ci: build helia-interop from ipfs/helia#1083

Temporary. The published @helia/interop still expects Kubo's old CIDv0 /
256 KiB import defaults and fails against the new unixfs-v1-2025 default,
so build it from the PR branch that adapts the suite (ipfs/helia#1083).
Revert this commit once that PR ships to npm.

lidel commented on .github/workflows/interop.yml:
```diff
@@ -73,53 +73,21 @@ jobs:
-        if: steps.helia-cache.outputs.cache-hit != 'true'
-        run: npm install @helia/interop
-      # Recurring @helia/interop regression: the published .aegir.js is unusable
-      # from inside node_modules, in one of two ways.
```
- [x] revert this if/once https://github.com/ipfs/helia/pull/1083 lands

guillaumemichel commented on core/coreapi/unixfs.go:
```diff
@@ -35,13 +37,100 @@ var log = logging.Logger("coreapi")
+	configCidVer := int(imp.CidVersion.WithDefault(config.LegacyFallbackCidVersion))
+	if !callerForcesVersion || caller.CidVersion == configCidVer {
+		if !imp.CidVersion.IsDefault() {
+			opts = append(opts, options.Unixfs.CidVersion(configCidVer))
```
IIUC `add` and MFS disagree when a `CidVersion=0` config meets a non-sha2-256 hash?

With `{"Import": {"CidVersion": 0, "HashFunction": "blake3"}}` (which `ValidateImportConfig` accepts), this emits `CidVersion(0)` and L85 emits `Hash(blake3)` — and `UnixfsAddOptions` rejects that pair outright:

https://github.com/ipfs/kubo/blob/13ad613c5884894da89f157bcdf11f4163f97401/core/coreiface/options/unixfs.go#L130-L131

Whereas `UnixFSCidBuilder` silently bumps v0 → v1 for the same config, so `ipfs files write` succeeds and emits CIDv1. Same repo, same config: `add` errors, MFS quietly changes version.

This inconsistency predates this PR, but it may be a good opportunity to address it?

lidel replied:
Good catch, and yes: `add` refused every import while MFS quietly bumped the version, on a config `ValidateImportConfig` accepted.

Fixed in 5b309d4ac by refusing the config instead of picking a winner: `Import.CidVersion=0` next to a non-sha2-256 `Import.HashFunction` cannot be honored as written, so the node now says so at startup. An unset `CidVersion` is untouched, the hash still selects CIDv1 there, and `add` and MFS agree on every repo that starts.

guillaumemichel commented on core/commands/add.go:
```diff
@@ -428,16 +393,27 @@ https://github.com/ipfs/kubo/blob/master/docs/config.md#import
+		if cidVersionSetByUser && cidVer == 0 && rawblks {
+			return fmt.Errorf("%s can't be true with --cid-version=0 (CIDv0 requires dag-pb leaves)", rawLeavesOptionName)
+		}
+
```
Wouldn't it be possible to have the root with CIDv0 and still pass `--raw-leaves`, encoding the leaves using CIDv1? I just want to make sure we don't break any existing use cases.

IIUC the sharness suite already asserts exactly that shape, in files this PR doesn't touch:

- `t0040-add-and-cat.sh:911` — `test_add_cat_5MB --raw-leaves "QmabWSFaPusmiZaaVZLhEUtHcj8CCvVeUfkBpKqAkKVMiS"`
- `t0040-add-and-cat.sh:860-863` — `add_directory '--raw-leaves'` expects `PLANETS="QmfWfQfKCY5Ukv9peBbxM5vqWM9BzmqUSXvdCgjT2wsiBT"` with `MARS`/`VENUS` as `bafkrei...`, i.e. a CIDv0 dag-pb root linking CIDv1 raw leaves
- `t0270-filestore.sh:16` — `EXPHASH="QmXKtATsEt42CF5JoSsmzJstrvwEB5P89YQtdX4mdf9E3M"` from `ipfs add --raw-leaves --nocopy`

And these two still reach that shape without tripping the check:

- `ipfs add --raw-leaves` on a `unixfs-v0-2015` repo — `cidVersionSetByUser` is false, so the guard is skipped
- `ipfs add --cid-version=0 --nocopy` — `--raw-leaves` is never passed so `rawblks` is false, and `UnixfsAddOptions` then auto-enables raw leaves for nocopy

Am I missing something about the intent? If the goal is to steer people toward `unixfs-v1-2025` rather than to forbid the shape outright, a warning might fit better than a hard error. Otherwise we could simply drop the check, the combination is well-defined and already exercised in the suite.

lidel replied:
You're right, the shape is valid. Dropped the check in 8494f9731, along with the config-level one, which was worse: it stopped the daemon and every offline command on repos that only set `Import.UnixFSRawLeaves=true`, the way to get raw leaves before the profiles existed.

Tests now guard both: the pairing stays accepted in `config/import_test.go`, and such a repo keeps producing its old CIDs in `test/cli/add_test.go`.

guillaumemichel commented on client/rpc/unixfs.go:
```diff
@@ -42,7 +42,6 @@ func (api *UnixfsAPI) Add(ctx context.Context, f files.Node, opts ...caopts.Unix
 	req := api.core().Request("add").
 		Option("hash", mht).
 		Option("chunker", options.Chunker).
```
IIUC `chunker` and `hash` are still forwarded unconditionally here, so the daemon's `Import` defaults never reach `client/rpc` callers?

L32's `caopts.UnixfsAddOptions(opts...)` materializes the library defaults, so even when the caller passes no options at all, `options.Chunker` is `"size-262144"` and `mht` is `"sha2-256"`, and both go on the wire regardless. But the daemon reads an empty value as "not set, use config":

https://github.com/ipfs/kubo/blob/13ad613c5884894da89f157bcdf11f4163f97401/core/commands/add.go#L487-L498

Since the request always carries a non-empty value, `!= ""` is always true there and the "otherwise `UnixfsAPI.Add` applies the Import config default" path never runs for RPC callers. So on a fresh `unixfs-v1-2025` repo (`UnixFSChunker = "size-1048576"`), `ipfs add` chunks at 1 MiB while `api.Unixfs().Add(ctx, f)` against that same daemon chunks at 256 KiB, leading to different CIDs for anything over 256 KiB.

Would adding `ChunkerSet`/`MhTypeSet` to `UnixfsAddSettings` and gating these two the same way as `CidVersionSet` just below be the right fix? `hash` looks harmless today since both profiles pin sha2-256, but a repo with a custom `Import.HashFunction` would hit the same issue.

lidel replied:
Exactly right, and that was the fix: 38277e602 adds `ChunkerSet`, `MhTypeSet` and `LayoutSet`, and sends each option only when the caller set it. Layout had the same hole in reverse, an explicit balanced request could not override a trickle config.

`client/rpc/unixfs_test.go` now compares RPC and `ipfs add` CIDs on the same daemon. It fails on the old code.

lidel committed:
Merge remote-tracking branch 'origin/master' into feat/puts-return-cidv1

# Conflicts:
#	.github/workflows/interop.yml

lidel committed:
fix: apply default import profile in daemon --init

`ipfs daemon --init` built its config through its own call to
config.InitWithIdentity, so a repo created that way skipped the
unixfs-v1-2025 profile, stayed on CIDv0 and printed the implicit-defaults
notice on the repo it had just written. Both init paths now share
NewGeneratedConfig.

lidel committed:
fix: honor explicit --trickle=false on ipfs add

Resolving the DAG layout in UnixfsAPI.Add dropped the flag's set/unset
state, so a repo with Import.UnixFSDAGLayout=trickle produced trickle
DAGs even when the command asked for balanced. The command now forwards
the layout it was given whenever --trickle appears.

lidel committed:
fix(rpc): send only the import options the caller set

client/rpc always sent its own hash, chunker and DAG layout with every
Unixfs().Add, which overrode the daemon's Import config: against a
unixfs-v1-2025 node the library chunked at 256 KiB while `ipfs add`
used 1 MiB, so the same bytes got two different CIDs. Track whether the
caller set each option and forward only those, matching how cid-version
and raw-leaves are already handled.

lidel committed:
fix: keep allowing raw leaves with CIDv0

A CIDv0 dag-pb root over raw (CIDv1) leaves is a valid DAG and is what
`ipfs add --nocopy` writes, so rejecting the combination broke existing
repos that set Import.UnixFSRawLeaves=true on the implicit CIDv0 default:
the daemon and every offline command refused to build a node. Drop the
config and CLI checks, and document the mixed DAG instead.

lidel committed:
chore(config): keep Default* import consts as deprecated aliases

Renaming them to LegacyFallback* broke every external program that
imports kubo's config package.

lidel committed:
docs: move CIDv1 default highlight to v0.44 changelog

v0.43 shipped while this branch was open. Also states the new default
once in the Import section of the config reference, instead of repeating
it per option.

lidel committed:
fix: refuse CIDv0 config with a non-sha2-256 hash

Import.CidVersion=0 next to a hash CIDv0 cannot carry made `ipfs add`
fail on every import while `ipfs files write` silently wrote CIDv1,
because UnixFSCidBuilder bumps the version and UnixfsAddOptions rejects
the pair. Refuse the config instead, so both agree on every repo that
starts. An unset CidVersion is unaffected: the hash selects CIDv1.

lidel committed:
fix(fuse): seed unpublished IPNS roots with the repo CID version

The IPNS mount starts a never-published key from boxo's empty directory
node, which is CIDv0. On a CIDv1 repo MFS re-encodes that root, so the
republisher saw a value it had never published and unmount waited out
its 5s close timeout publishing an empty directory. Daemon shutdown with
FUSE mounted went from 0.1s to 5.1s, past the test harness SIGTERM
budget, which then killed the daemon and left stale mounts behind.

lidel committed:
ci: keep apt from eating the test job budget

unit-tests, cli-tests and sharness-test have a 15 minute budget and were
spending up to 13 of it installing packages, then dying to the job
timeout with most tests unrun. The runner's Ubuntu mirror has served as
slowly as 30 kB/s, so keep the packages in the workflow cache and let
apt install them from disk. Skip apt entirely when the tools are already
present, drop recommends (fish alone pulls in lynx), and cap the step so
a bad mirror fails under its own name.
